server/storage.go

```diff
@@ -561,3 +564,115 @@ func saveGDriveToken(path string, token *oauth2.Token, logger *log.Logger) {
 
 	json.NewEncoder(f).Encode(token)
 }
+
+type StorjStorage struct {
+	Storage
+	project *uplink.Project
+	bucket  *uplink.Bucket
+	logger  *log.Logger
+}
+
+func NewStorjStorage(access, bucket string, logger *log.Logger) (*StorjStorage, error) {
+	var instance StorjStorage
+	var err error
+
+	ctx := context.TODO()
+
+	parsedAccess, err := uplink.ParseAccess(access)
+	if err != nil {
+		return nil, err
+	}
+
+	instance.project, err = uplink.OpenProject(ctx, parsedAccess)
+	if err != nil {
+		return nil, err
+	}
+
+	instance.bucket, err = instance.project.EnsureBucket(ctx, bucket)
+	if err != nil {
+		return nil, err
+	}
+
+	instance.logger = logger
+
+	return &instance, nil
+}
+
+func (s *StorjStorage) Type() string {
+	return "storj"
+}
+
+func (s *StorjStorage) Head(token string, filename string) (contentType string, contentLength uint64, err error) {
+	key := storj.JoinPaths(token, filename)
+
+	ctx := context.TODO()
+
+	obj, err := s.project.StatObject(ctx, s.bucket.Name, key)
+	if err != nil {
+		return "", 0, err
+	}
+	contentType = obj.Standard.ContentType
+	contentLength = uint64(obj.Standard.ContentLength)
+
+	return
+}
+
+func (s *StorjStorage) Get(token string, filename string) (reader io.ReadCloser, contentType string, contentLength uint64, err error) {
+	key := storj.JoinPaths(token, filename)
+
+	s.logger.Printf("Getting file %s from Storj Bucket", filename)
+
+	ctx := context.TODO()
+
+	download, err := s.project.DownloadObject(ctx, s.bucket.Name, key, nil)
+	if err != nil {
+		return nil, "", 0, err
+	}
+	contentType = download.Info().Standard.ContentType
+	contentLength = uint64(download.Info().Standard.ContentLength)
+	//return the download as a reader
+	reader = download
+	return
+}
+
+func (s *StorjStorage) Delete(token string, filename string) (err error) {
+	key := storj.JoinPaths(token, filename)
+
+	s.logger.Printf("Deleting file %s from Storj Bucket", filename)
+
+	ctx := context.TODO()
+
+	_, err = s.project.DeleteObject(ctx, s.bucket.Name, key)
+
+	return
+}
+
+func (s *StorjStorage) Put(token string, filename string, reader io.Reader, contentType string, contentLength uint64) (err error) {
+	key := storj.JoinPaths(token, filename)
+
+	s.logger.Printf("Uploading file %s to Storj Bucket", filename)
+
+	ctx := context.TODO()
+
+	writer, err := s.project.UploadObject(ctx, s.bucket.Name, key, &uplink.UploadOptions{})
+	if err != nil {
+		return err
+	}
+
+	n, err := io.Copy(writer, reader)
+	if err != nil {
+		return err
```

Comment:
call writer.Abort, since the reader might have errored and you don't want:
1. leave stuff lying around on the satellite
2. leave connections open

Reply:
done